backend/de.metas.monitoring/src/main/java/de/metas/monitoring/adapter/MicrometerPerformanceMonitoringService.java

```diff
@@ -42,63 +42,88 @@
+		final PerformanceMonitoringData perfMonData = perfMonDataTL.get();
+		final ArrayList<Tag> tags = new ArrayList<>();
+
+		mkTagIfNotNull("name", metadata.getClassName()).ifPresent(tags::add);
+		mkTagIfNotNull("action", metadata.getFunctionName()).ifPresent(tags::add);
+
+		if(perfMonData.getDepth() == 0
+				&& metadata.getType() == Type.REST_CONTROLLER_WITH_WINDOW_ID
+				&& metadata.getFunctionName() != null
+				&& metadata.getWindowNameAndId() != null)
 		{
-			callableToUse = () -> apmPerformanceMonitoringService.get().monitorTransaction(callable, metadata);
+			perfMonData.setIsInitiatorLabelActive(true);
+			perfMonData.setInitiator(metadata.getClassName() + " - " + metadata.getFunctionName());
+			perfMonData.setInitiatorWindow(metadata.getWindowNameAndId());
+			perfMonData.getCalledBy().add(0, "HTTP Request");
+
 		}
-		else
+		else if(perfMonData.getDepth() == 0
+				&& metadata.getType() == Type.REST_CONTROLLER
+				&& metadata.getFunctionName() != null)
 		{
-			callableToUse = callable;
+			perfMonData.setIsInitiatorLabelActive(true);
+			perfMonData.setInitiator(metadata.getClassName() + " - " + metadata.getFunctionName());
+			perfMonData.setInitiatorWindow("NONE");
+			perfMonData.getCalledBy().add(0, "HTTP Request");
 		}
-		return recordCallable(callableToUse, tags, "mf." + metadata.getType().getCode());
-	}
-
-	@Override
-	public <V> V monitorSpan(
-			@NonNull final Callable<V> callable,
-			@NonNull final SpanMetadata metadata)
-	{
-		final ArrayList<Tag> tags = createTagsFromLabels(metadata.getLabels());
-
-		mkTagIfNotNull("Name", metadata.getName()).ifPresent(tags::add);
-		mkTagIfNotNull("SubType", metadata.getSubType()).ifPresent(tags::add);
-		mkTagIfNotNull("Action", metadata.getAction()).ifPresent(tags::add);
-
-		final Callable<V> callableToUse;
-		if (apmPerformanceMonitoringService.isPresent())
+		else if(perfMonData.getDepth() == 0)
 		{
-			callableToUse = () -> apmPerformanceMonitoringService.get().monitorSpan(callable, metadata);
+			perfMonData.setIsInitiatorLabelActive(false);
 		}
-		else
+
+		if(perfMonData.getIsInitiatorLabelActive())
 		{
-			callableToUse = callable;
+			mkTagIfNotNull("depth", String.valueOf(perfMonData.getDepth())).ifPresent(tags::add);
+			mkTagIfNotNull("initiator", perfMonData.getInitiator()).ifPresent(tags::add);
+			mkTagIfNotNull("window", perfMonData.getInitiatorWindow()).ifPresent(tags::add);
+			mkTagIfNotNull("callerName", metadata.getClassName() + " - " + metadata.getFunctionName()).ifPresent(tags::add);
+			mkTagIfNotNull("calledBy", perfMonData.getCalledBy().get(perfMonData.getDepth())).ifPresent(tags::add);
+			perfMonData.getCalledBy().add(perfMonData.getDepth() + 1 ,metadata.getClassName() + " - " + metadata.getFunctionName());
+
+			final Type type;
+			if(isAnyRestControllerType(metadata))
+			{
+				type = metadata.getType();
+			}
+			else
+			{
+				type = Type.REST_API_PROCESSING;
+			}
+
+			perfMonData.setDepth(perfMonData.getDepth() + 1);
```

Comment:
why not introduce a method called incrementDepth() instead of setDepth.
same, decrementDepth() if needed